### PR TITLE
fix: delete existing RBACs when namespace label is updated

### PR DESCRIPTION
**Pull request by shubhamagarwal19**
**What type of PR is this?**
/kind bug

**What does this PR do / why we need it**:
This PR fixes the issue when namespace label `argocd.argoproj.io/managed-by` is updated to a different value i.e from namespace `foo` -> `bar`. 
The issue being, RBACs that were created for "foo" does not get removed when the namespace label is updated, while it creates the RBACs for the new namespace (bar). 

**Have you updated the necessary documentation?**

* [ ] Documentation update is required by this PR.
* [ ] Documentation has been updated.

**Which issue(s) this PR fixes**:

Fixes #?
https://issues.redhat.com/browse/GITOPS-1180

**How to test changes / Special notes to the reviewer**:
- Deploy the latest Operator.
- deploy two argocd instances in the namespace `foo1` and `foo2`.
- create a namesapce `bar` with the label `argocd-argoproj.io/managed-by: foo1`.
- Make sure argocd instance in `foo1` can manage the resources in `bar` namespace.
- update the namespace label value from `foo1` -> `foo2`
- Make sure RBACs for `foo2` are created and RBACs for `foo1` gets deleted.

### CODE REVIEW

File: pkg/controller/argocd/util.go [hunk 1]
```diff
@@ -1097,10 +1097,28 @@ func containsString(arr []string, s string) bool {
...
-			// Event is then handled by the reconciler.
-			if _, ok := e.MetaNew.GetLabels()[common.ArgoCDManagedByLabel]; ok {
-				return true
+			// This checks if ArgoCDManagedByLabel exists in newMeta, if exists then -
+			// 1. Check if oldMeta had the label or not? if no, return true
+			// 2. if yes, check if old and new value are the same, if yes, return true.
+			// 3. If old and new value are not the same, first deleteRBACs for old val, return true.
+			// Event is then handled by the reconciler, which would create appropriate RBACs.
+			if valNew, ok := e.MetaNew.GetLabels()[common.ArgoCDManagedByLabel]; ok {
+				if valOld, ok := e.MetaOld.GetLabels()[common.ArgoCDManagedByLabel]; !ok || valOld == "" {
```
Comment (iam-veeramalla) [inline pkg/controller/argocd/util.go:1106]: Can this if , if-else and else block be modified into Just one if block ?

if valOld != valNew {
      // If old and new value are not the same, first deleteRBACs for old val, return true.
     // Event is then handled by the reconciler, which would create appropriate RBACs.
}

Comment (shubhamagarwal19) [reply pkg/controller/argocd/util.go]: Thanks @iam-veeramalla, fixed it!!

**Comment by jannfis** [comment]
Just a question: Creation of the new RBAC roles is handled somewhere down the road after deleting the old ones in case of update, right?

**Comment by shubhamagarwal19** [comment]
> Just a question: Creation of the new RBAC roles is handled somewhere down the road after deleting the old ones in case of update, right?

Yep! after it's done deleting the RBACs for the older namespace, it would create them for the updated namespace value.

File: pkg/controller/argocd/util.go [hunk 2]
```diff
@@ -1097,9 +1097,23 @@ func containsString(arr []string, s string) bool {
...
 	return predicate.Funcs{
 		UpdateFunc: func(e event.UpdateEvent) bool {
-			// If ArgoCDManagedByLabel exists, return true.
-			// Event is then handled by the reconciler.
-			if _, ok := e.MetaNew.GetLabels()[common.ArgoCDManagedByLabel]; ok {
+			// This checks if ArgoCDManagedByLabel exists in newMeta, if exists then -
+			// 1. Check if oldMeta had the label or not? if no, return true
+			// 2. if yes, check if the old and new values are different, if yes,
+			// first deleteRBACs for the old value & return true.
+			// Event is then handled by the reconciler, which would create appropriate RBACs.
```
Comment (jannfis) [inline pkg/controller/argocd/util.go:1104]: :+1: thanks for documenting!

**Review by jannfis — APPROVED** [review]
LGTM

File: pkg/controller/argocd/util.go [hunk 3]
```diff
@@ -1097,9 +1097,23 @@ func containsString(arr []string, s string) bool {
...
+			if valNew, ok := e.MetaNew.GetLabels()[common.ArgoCDManagedByLabel]; ok {
+				if valOld, ok := e.MetaOld.GetLabels()[common.ArgoCDManagedByLabel]; ok && valOld != valNew {
+					k8sClient, err := initK8sClient()
+					if err != nil {
+						return false
+					}
+					if err := deleteRBACsForNamespace(valOld, e.MetaOld.GetName(), k8sClient); err != nil {
```
Comment (wtam2018) [inline pkg/controller/argocd/util.go:1111]: Deleting RBAC inside of Predicate Func does not sound right.  Predicate Func is used to filter event.

**Pull request merged**